src/content/blog/val-vibes.mdx

```diff
@@ -0,0 +1,141 @@
+import Search1 from "./val-vibes/search1.png";
+import Search2 from "./val-vibes/search2.png";
+import Blobindex from "./val-vibes/blobindex.png";
+import Blobsearch from "./val-vibes/blobsearch.png";
+import ManLicksDog from "./val-vibes/man-licks-dog.jpg";
+
+### Similar vibes
+
+“Search” is finding documents based on a query, using some sort of sorting mechanism that ranks the most relevant documents. One such sorting mechanism is “semantic search”, which uses one of those crazy features to come out of the world of machine learning: embeddings. It works like this: you put in a string into an API, and you get an array of numbers (the “embedding”), with the property that things that have the same “vibe” will get similar numbers.
+
+For example, “animal that barks” should have similar vibes to “dog”. And “dog bites man” should have different vibes than “man bites dog”.
+
+<figure style={{ maxWidth: "300px" }}>
+  <Image src={ManBitesDog} alt="Man bites dog" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+You generate embeddings for all your documents (ahead of time), as well as for the search query, and then you return documents sorted by how similar the embeddings are.
+
+That's the idea, though in practice embeddings can be quite unpredictable and surprising. Our examples above already break down… “animal that barks” apparently vibes more with “woof woof woof” than with “dog”. And “dog bites man” and “man bites dog” have even more similar vibes than that, presumably because there's lots of biting going on in both cases.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={CompareEmbeddings} alt="" />
+  <figcaption>
+    Woof woof woof.
+  </figcaption>
+</figure>
+
+At Val Town we had existing search for executable code snippets (“vals”), but it was quite bad. It did only exact matches, often returning only a few results, or none at all. Surely semantic search would help return vals with similar vibes, despite getting confused by barking and biting.
+
+### Prototyping in Val Town
+
+I started prototyping this on my local computer, but then realized: this is Val Town, the ultimate prototyping tool for things like this. Someone had already made an <a href="https://www.val.town/v/sqlite/db">excellent val</a> for querying all public vals:
+
+<figure style={{ maxWidth: "600px" }}>
+  <Image src={QueryVals} alt="" />
+  <figcaption>
+    Querying public vals, by <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a>.
+  </figcaption>
+</figure>
+
+I ended up making three different prototypes, using different storage engines: [Postgres](https://www.postgresql.org/) (using [Neon](https://neon.tech/)), [SQLite](https://sqlite.org/) (using [Turso](https://turso.tech/)), and [blob storage](https://docs.val.town/std/blob/) (using [Cloudflare R2](https://www.cloudflare.com/developer-platform/r2/)). Let's look at Postgres first, even though I built it last, since it's the simplest. At the end I'll also show how to do this with blob storage.
```

Comment:
In some sense, this is the lede ("I made three different vector search prototypes for vals, in val town. Let me walk you through em.") and it's a bit buried down here

I think I might rebrand "blob storage" as "in-memory" or "manually in javascript"

Reply:
You could also throw Turso a shoutout for their new vector search that wasn't ready in time for this blog post

Reply:
The folks at turso might do the legwork involved in porting these examples over to their new tech if we told them how... I can intro you or you can let them know in the discord channel